qualtran/bloqs/chemistry/prepare_mps/decompose_gate_hr.py

```diff
@@ -0,0 +1,240 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import Dict, List, Tuple
+
+import attrs
+import numpy as np
+
+from qualtran import Bloq, BloqBuilder, BoundedQUInt, Register, Signature, SoquetT
+from qualtran.bloqs.basic_gates import CNOT, Hadamard, ZGate
+from qualtran.bloqs.basic_gates.x_basis import XGate
+from qualtran.bloqs.reflection_using_prepare import ReflectionUsingPrepare
+from qualtran.bloqs.rotations.phase_gradient import PhaseGradientState
+from qualtran.bloqs.select_and_prepare import PrepareOracle
+from qualtran.bloqs.state_preparation.state_preparation_via_rotation import (
+    StatePreparationViaRotations,
+)
+
+
+@attrs.frozen
+class DecomposeGateViaHR(Bloq):
+    r"""Given all or some columns of an unitary, it is generated via Householder reflections.
+
+    This bloq implements the algorithm described in [1], which basically implements an arbitrary
+    unitary operator $U$ with a reflection ancilla as
+
+    $$
+        |0\rangle\langle 1| U + |1\rangle\langle 0| U^\dagger$,
+    $$
+
+    even though for the purpose of making this bloq simpler to work with, the gate actually
+    implemented is
+
+    $$
+        |0\rangle\langle 1| U + |1\rangle\langle 0| U^\dagger$.
+    $$
+
+    For a detailed description of how to use this bloq refer to its tutorial.
+
+    Args:
+        phase_bitsize: size of the register that is used to store the rotation angles. Bigger values
+            increase the accuracy of the results.
+        gate_cols: tuple that contains, for each entry, a pair where the first element is the
+            index of the column and the second a tuple that contains the complex values of each
+            column. For example, for the gate [[a,b,],[c,d]] it would be ((0,(a,c)), (1(b,d))).
+        uncompute: wether to implement U or U^t. Either case the gate_cols to be provided are those
+            of U.
+        internal_phase_grad: a phase gradient state is needed for the decomposition. It can be
+            either be provided externally if this attribute is set to False or internally otherwise.
+        internal_refl_ancilla: this bloq also uses a reflection ancilla that can be provided
+            externally if this attribute is set to false or internally otherwise. In the case of the
+            internal use it will be projected into |0>, and all the data that was contained into the
+            |1> projection will be lost.
+
+    References:
+        [Synthesis of unitaries with Clifford+T circuits]
+        (https://arxiv.org/pdf/1306.3200.pdf).
+            Kliuchnikov 2013.
+    """
+    phase_bitsize: int  # number of ancilla qubits used to encode the state preparation's rotations
+    gate_cols: Tuple[
+        int, Tuple[complex, ...]
+    ]  # tuple with the columns of the gate that are specified
```

Comment:
so I think this is actually `Tuple[Tuple[int, Tuple[complex, ...], ...]`.

A less specific but (imo) more readable way of writing it would be `Sequence[Tuple[int, Sequence[complex]]]`